llama.cpp

```diff
@@ -32,6 +32,12 @@
 #endif
 #endif
 
+// TODO: Fix unused logit skipping crashes on ROCm
+// (see https://github.com/ggerganov/llama.cpp/pull/2700#issuecomment-1689548127)
+#ifndef LLAMA_USE_HIPBLAS
+#define LLAMA_SKIP_UNUSED_LOGITS
+#endif
+
```

Comment:
I suggest removing this and instead make LLAMA_SKIP_UNUSED_LOGITS configurable from make/CMake.

Reply:
Guess it depends on whether this works for everything _except_ ROCM.

Reply:
Yes I was hoping for this to be a default mini optimization that always works *except* in some temporary cases, but I'll prepare a cmake / makefile update.

I've tested the output on CUDA again ([updated colab](https://colab.research.google.com/gist/ochafik/664f78524d8650a7b71b322d1b55d98d/llama-cpp-skip-unused-logits-computations.ipynb#scrollTo=aU88u6VBTQwX&line=2&uniqifier=1)), included a branch where I've merged the ROCm changes and the CUDA backend appears to skip well on T4 w/ and w/o the ROCm changes.

Reply:
> included a branch where I've merged the ROCm changes and the CUDA backend appears to skip well on T4 w/ and w/o the ROCm changes.

I'm not sure I fully understand this. Do you mean you're saying the changes like:

```patch
- const float dall = x[i].dm.x;
+ const float dall = __low2half(x[i].dm);
```

aren't the issue when running with CUDA? That's certainly useful information since it would mean the non-ROCM-specific changes don't cause a problem for CUDA.

Most likely the problem I originally described is specific to actually using ROCM on an AMD GPU.